feat: remove unused assets css in dist

luffy84217 opened the pull request:
### Description

When one imports an asset (let's say .svg) using import.meta.glob(..., { eager: true, import: 'default' }) new path will be inlined in a const and then this const get's tree-shaked/optimized, corresponding asset is not excluded from the final build.

It's expected that no unused assets, especially heavy ones like fonts and images, are not included in final build.

Fix: https://github.com/vitejs/vite/issues/19949
Ref: https://github.com/BerkliumBirb/vite-broken-tree-shaking

<!-- What is this PR solving? Write a clear description or reference the issues it solves (e.g. `fixes #123`). What other alternatives have you explored? Are there any parts you think require more attention from reviewers? -->

<!----------------------------------------------------------------------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md.
- Check that there isn't already a PR that solves the problem the same way. If you find a duplicate, please help us reviewing it.
- Update the corresponding documentation if needed.
- Include relevant tests that fail without this PR but pass with it.

Thank you for contributing to Vite!
----------------------------------------------------------------------->

sapphi-red commented:
> @sapphi-red Hi, since this is my very first PR to vite, if I'm lack of consideration, please give me advice. Since you changed the type from bug to feature, does it need test for it?
> https://github.com/vitejs/vite/issues/19949#issuecomment-3292380571

@luffy84217 The tests are needed for both bugs and features.

sapphi-red reviewed (changes requested):
The tests are failing.

Also the code doesn't look right. This would remove all assets that are not used from chunks and css files. But the asset might be emitted by a plugin that Vite itself is not aware of and Vite shouldn't remove those assets.
I guess we can use [`needsCodeReference`](https://rollupjs.org/plugin-development/#:~:text=If-,needsCodeReference,-is%20set%20to) if we can mark the asset as used. I'm not sure if we can mark the asset as used though.

luffy84217 committed:
feat: remove unused assets css in dist

When one imports an asset (let's say .svg) using import.meta.glob(..., {
eager: true, import: 'default' }) new path will be inlined in a const
and then this const get's tree-shaked/optimized, corresponding asset is
not excluded from the final build.

It's expected that no unused assets, especially heavy ones like fonts
and images, are not included in final build.

Fix: https://github.com/vitejs/vite/issues/19949
Ref: https://github.com/BerkliumBirb/vite-broken-tree-shaking

luffy84217 committed:
Update packages/vite/src/node/plugins/asset.ts

Co-authored-by: graphite-app[bot] <96075541+graphite-app[bot]@users.noreply.github.com>

sapphi-red commented:
Yes, it requires some investigation and a medium-size change. Understood, I’ll go ahead and close this PR. Thanks!
